File: internal/db/sqladapter/partition.go
package sqladapter

import (
	"regexp"
	"strings"
)

// SHOW CREATE TABLE DDL clauses we care about for AI/query hints:
//
//	PARTITION BY RANGE(`event_date`)                 (Doris/MySQL)
//	PARTITION BY LIST(`region`)
//	PARTITION BY RANGE (to_days(`d`))                (MySQL expression)
//	DUPLICATE KEY(`id`, `event_date`)                (Doris)
//	UNIQUE KEY(`id`) / AGGREGATE KEY(`id`)           (Doris)
//	PRIMARY KEY (`id`)                               (MySQL/Doris)
var (
	partitionByRe = regexp.MustCompile(`(?is)PARTITION\s+BY\s+([A-Z]+)?(?:\s+COLUMNS)?\s*\(`)
	keyClauseRe   = regexp.MustCompile(`(?is)(DUPLICATE|UNIQUE|AGGREGATE|PRIMARY)\s+KEY\s*\(`)
)

// parseCreateTableShape extracts a table's partition and key columns from its
// SHOW CREATE TABLE DDL, as compact human/AI-readable strings. Either may be
// empty when the table has no partitioning or no declared key. Results are used
// both to enrich AI schema context and to display in the table metadata view.
//
// partition: e.g. "RANGE(event_date)" or "LIST(region)" or "(event_date)".
// key:       e.g. "DUPLICATE KEY(id, event_date)" or "PRIMARY KEY(id)".
func parseCreateTableShape(ddl string) (partition, key string) {
	return parsePartitionClause(ddl), parseKeyClause(ddl)
}

func parsePartitionClause(ddl string) string {
	loc := partitionByRe.FindStringSubmatchIndex(ddl)
	if loc == nil {
		return ""
	}
	// Partition method (RANGE/LIST/HASH/…) if present; loc[2:4] is group 1.
	method := ""
	if loc[2] >= 0 {
		method = strings.ToUpper(strings.TrimSpace(ddl[loc[2]:loc[3]]))
	}
	// The matched "(" is the last byte of the overall match (loc[1]-1).
	inner, ok := balancedParen(ddl, loc[1]-1)
	if !ok {
		return ""
	}
	cols := columnIdentifiers(inner)
	if len(cols) == 0 {
		return ""
	}
	joined := strings.Join(cols, ", ")
	if method == "" {
		return "(" + joined + ")"
	}
	return method + "(" + joined + ")"
}

func parseKeyClause(ddl string) string {
	loc := keyClauseRe.FindStringSubmatchIndex(ddl)
	if loc == nil {
		return ""
	}
	kind := strings.ToUpper(strings.TrimSpace(ddl[loc[2]:loc[3]]))
	inner, ok := balancedParen(ddl, loc[1]-1)
	if !ok {
		return ""
	}
	cols := columnIdentifiers(inner)
	if len(cols) == 0 {
		return ""
	}
	return kind + " KEY(" + strings.Join(cols, ", ") + ")"
}

// balancedParen returns the substring inside the parentheses that begins at
// openIdx (which must point at '('), honoring nesting so an expression like
// to_days(`d`) is captured whole. ok is false when the parenthesis is unbalanced.
func balancedParen(s string, openIdx int) (string, bool) {
	if openIdx < 0 || openIdx >= len(s) || s[openIdx] != '(' {
		return "", false
	}
	depth := 0
	for i := openIdx; i < len(s); i++ {
		switch s[i] {
		case '(':
			depth++
		case ')':
			depth--
			if depth == 0 {
				return s[openIdx+1 : i], true
			}
		}
	}
	return "", false
}

var identifierRe = regexp.MustCompile("[`\"]?([A-Za-z_][A-Za-z0-9_]*)[`\"]?")

// columnIdentifiers pulls the bare column identifiers out of a partition/key
// column list, dropping function names and SQL keywords (e.g. to_days(`d`) →
// "d"). It splits on commas at the top level so multi-column lists keep order.
func columnIdentifiers(inner string) []string {
	var cols []string
	seen := map[string]bool{}
	for _, part := range splitTopLevel(inner) {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		// For an expression like to_days(`d`) take the inner-most identifier; for
		// a plain `col` it is the identifier itself. Pick the last identifier match
		// (the column, not the function name).
		matches := identifierRe.FindAllStringSubmatch(part, -1)
		if len(matches) == 0 {
			continue
		}
		name := matches[len(matches)-1][1]
		if name == "" || isSQLNoise(name) || seen[strings.ToLower(name)] {
			continue
		}
		seen[strings.ToLower(name)] = true
		cols = append(cols, name)
	}
	return cols
}

// splitTopLevel splits on commas that are not inside nested parentheses.
func splitTopLevel(s string) []string {
	var out []string
	depth, start := 0, 0
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '(':
			depth++
		case ')':
			depth--
		case ',':
			if depth == 0 {
				out = append(out, s[start:i])
				start = i + 1
			}
		}
	}
	out = append(out, s[start:])
	return out
}

// isSQLNoise drops common function/keyword tokens that are not column names.
func isSQLNoise(name string) bool {
	switch strings.ToLower(name) {
	case "asc", "desc":
		return true
	}
	return false
}

File: src/db/sql/partition.rs
//! Parse partition / key clauses out of `SHOW CREATE TABLE` DDL. Port of
//! `internal/db/sqladapter/partition.go`.
//!
//! Examples produced:
//!   partition: "RANGE(event_date)" / "LIST(region)" / "(event_date)"
//!   key:       "DUPLICATE KEY(id, event_date)" / "PRIMARY KEY(id)"

use once_cell::sync::Lazy;
use regex::Regex;

static PARTITION_BY_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?is)PARTITION\s+BY\s+([A-Z]+)?(?:\s+COLUMNS)?\s*\(").unwrap());
static KEY_CLAUSE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?is)(DUPLICATE|UNIQUE|AGGREGATE|PRIMARY)\s+KEY\s*\(").unwrap());
static IDENTIFIER_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"[`"]?([A-Za-z_][A-Za-z0-9_]*)[`"]?"#).unwrap());

/// Extract `(partition, key)` from DDL; either may be empty.
pub fn parse_create_table_shape(ddl: &str) -> (String, String) {
    (parse_partition_clause(ddl), parse_key_clause(ddl))
}

fn parse_partition_clause(ddl: &str) -> String {
    let Some(caps) = PARTITION_BY_RE.captures(ddl) else {
        return String::new();
    };
    let whole = caps.get(0).unwrap();
    let method = caps
        .get(1)
        .map(|m| m.as_str().trim().to_uppercase())
        .unwrap_or_default();
    // The matched "(" is the last byte of the overall match.
    let open_idx = whole.end() - 1;
    let Some(inner) = balanced_paren(ddl, open_idx) else {
        return String::new();
    };
    let cols = column_identifiers(inner);
    if cols.is_empty() {
        return String::new();
    }
    let joined = cols.join(", ");
    if method.is_empty() {
        format!("({joined})")
    } else {
        format!("{method}({joined})")
    }
}

fn parse_key_clause(ddl: &str) -> String {
    let Some(caps) = KEY_CLAUSE_RE.captures(ddl) else {
        return String::new();
    };
    let kind = caps.get(1).unwrap().as_str().trim().to_uppercase();
    let open_idx = caps.get(0).unwrap().end() - 1;
    let Some(inner) = balanced_paren(ddl, open_idx) else {
        return String::new();
    };
    let cols = column_identifiers(inner);
    if cols.is_empty() {
        return String::new();
    }
    format!("{kind} KEY({})", cols.join(", "))
}

/// Substring inside the parentheses beginning at `open_idx` (must point at
/// '('), honoring nesting. Returns None on imbalance.
fn balanced_paren(s: &str, open_idx: usize) -> Option<&str> {
    let bytes = s.as_bytes();
    if open_idx >= bytes.len() || bytes[open_idx] != b'(' {
        return None;
    }
    let mut depth = 0i32;
    let mut i = open_idx;
    while i < bytes.len() {
        match bytes[i] {
            b'(' => depth += 1,
            b')' => {
                depth -= 1;
                if depth == 0 {
                    return Some(&s[open_idx + 1..i]);
                }
            }
            _ => {}
        }
        i += 1;
    }
    None
}

/// Bare column identifiers from a partition/key column list, dropping function
/// names and noise keywords (`to_days(\`d\`)` → "d"), preserving order, deduped.
fn column_identifiers(inner: &str) -> Vec<String> {
    let mut cols = Vec::new();
    let mut seen = std::collections::HashSet::new();
    for part in split_top_level(inner) {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        let matches: Vec<_> = IDENTIFIER_RE.captures_iter(part).collect();
        let Some(last) = matches.last() else {
            continue;
        };
        let name = last.get(1).unwrap().as_str();
        let lower = name.to_lowercase();
        if name.is_empty() || is_sql_noise(&lower) || seen.contains(&lower) {
            continue;
        }
        seen.insert(lower);
        cols.push(name.to_string());
    }
    cols
}

/// Split on commas not inside nested parentheses.
fn split_top_level(s: &str) -> Vec<&str> {
    let bytes = s.as_bytes();
    let mut out = Vec::new();
    let mut depth = 0i32;
    let mut start = 0usize;
    for i in 0..bytes.len() {
        match bytes[i] {
            b'(' => depth += 1,
            b')' => depth -= 1,
            b',' if depth == 0 => {
                out.push(&s[start..i]);
                start = i + 1;
            }
            _ => {}
        }
    }
    out.push(&s[start..]);
    out
}

fn is_sql_noise(lower: &str) -> bool {
    matches!(lower, "asc" | "desc")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_range_partition_and_duplicate_key() {
        let ddl = "CREATE TABLE `t` (...) ENGINE=OLAP\nDUPLICATE KEY(`id`, `event_date`)\nPARTITION BY RANGE(`event_date`) ()";
        let (p, k) = parse_create_table_shape(ddl);
        assert_eq!(p, "RANGE(event_date)");
        assert_eq!(k, "DUPLICATE KEY(id, event_date)");
    }

    #[test]
    fn parses_expression_partition_to_inner_identifier() {
        let ddl = "CREATE TABLE x (...) PARTITION BY RANGE (to_days(`d`)) ()";
        let (p, _) = parse_create_table_shape(ddl);
        assert_eq!(p, "RANGE(d)");
    }

    #[test]
    fn parses_primary_key() {
        let (_, k) = parse_create_table_shape("CREATE TABLE x (id int, PRIMARY KEY (`id`))");
        assert_eq!(k, "PRIMARY KEY(id)");
    }

    #[test]
    fn no_partition_no_key() {
        let (p, k) = parse_create_table_shape("CREATE TABLE x (id int)");
        assert!(p.is_empty());
        assert!(k.is_empty());
    }
}
